contracts/NFTVesting/IntervalVestingNFTTimeLock.sol

```diff
@@ -0,0 +1,230 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts (last updated v4.5.0) (token/ERC20/utils/TokenTimelock.sol)
+// SpartanLabs Contracts (NFTVesting)
+
+pragma solidity ^0.8.0;
+import "./IERC721.sol";
+
+/**
+ * @dev A single NFT holder contract that will allow a beneficiary to extract the
+ * NFT after a given vesting start time with a discount sent to the beneficiary based on
+ * the vesting duration of the NFT.
+ *
+ * On every interval epoch, the discount accrued by the locker is based off a set amount.
+ *
+ * Note that in order for discount in ETH to be valid, ETH must first be sent to this contract upon token locking.
+ */
+contract IntervalVestingNFTTimeLock {
+    // ERC721 basic token smart contract
+    IERC721 private immutable _nft;
+
+    // ERC721 basic token ID of contract being held
+    uint256 private immutable _tokenId;
+
+    // beneficiary of token after they are released
+    address private immutable _beneficiary;
+
+    // timestamp when token release is enabled
+    uint256 private immutable _vestingStartTime;
+
+    // Max discount allowed for a token in percentage
+    uint256 private immutable _maxDiscountPercentage;
+
+    // Max number of Interval for vesting
+    uint256 private immutable _maxIntervals;
+
+    // Duration for each Interval
+    uint256 private immutable _intervalDuration;
+
+    // Discount for every interval passed
+    uint256 private immutable _discountPerInterval;
+
+    // Events
+    event EthReceived(address indexed sender, uint256 amount);
+
+    modifier validRelease() {
+        require(
+            block.timestamp >= _vestingStartTime + _intervalDuration,
+            "Vesting Schedule is not Up yet."
+        );
+        _;
+    }
+
+    /**
+     * @dev Deploys a timelock instance that is able to hold the token specified, and will only release it to
+     * `beneficiary_` when {release} is invoked after `vestingStartTime_`. The vesting start time is specified as a Unix timestamp
+     * (in seconds).
+     *
+     *  For every set of duration passed after the vesting start time, the number of intervals would increase.
+     *  The discount will then be applied to the beneficiary according to number of intervals the token has been vested for.
+     *
+     *  The developer would have to send ETH to this contract on contract deployement for discount to be applied
+     */
+    constructor(
+        IERC721 nft_,
+        uint256 tokenId_,
+        address beneficiary_,
+        uint256 vestingStartTime_,
+        uint256 maxDiscountPercentage_,
+        uint256 maxInterval_,
+        uint256 intervalDuration_,
+        uint256 discountPerInterval_
+    ) {
+        require(
+            vestingStartTime_ > block.timestamp,
+            "TimeLock: vesting start time is before current time"
+        );
+        require(
```

Comment:
Considering that we already have max intervals, interval duration and discount per interval, do we still need maxDiscountPercentage? Seems to complicate things more than necessary for intervals